cli/install/install.go

```diff
@@ -375,43 +402,111 @@ func copyBuildedTT(binDir, path, version string, installCtx InstallCtx,
 	return err
 }
 
-// installTt installs selected version of tt.
-func installTt(binDir string, installCtx InstallCtx, distfiles string) error {
-	versions, err := getVersionsFromRepo(installCtx.Local, distfiles, "tt", search.GitRepoTT)
+// checkCommit checks the existence of a commit by hash.
+func checkCommit(input string, programName string, installCtx InstallCtx,
+	distfiles string) (string, string, error) {
+	pullRequestHash := ""
+	isPullRequest, _ := util.IsPullRequest(input)
+	if isPullRequest {
+		pullRequestHash, err := getCommit(installCtx.Local, distfiles,
+			programName, input)
+		if err != nil {
+			return input, pullRequestHash,
+				fmt.Errorf("%s: unable to get pull-request info", input)
+		}
+		pullRequestHash = pullRequestHash[0:util.MinCommitHashLength]
+		return input, pullRequestHash, nil
+	}
+
+	isRightFormat, err := util.IsValidCommitHash(input)
 	if err != nil {
-		return err
+		return input, pullRequestHash, err
 	}
 
+	if !isRightFormat {
+		return input, pullRequestHash, fmt.Errorf("hash has a wrong format")
+	}
+
+	returnedHash, err := getCommit(installCtx.Local, distfiles, programName, input)
+	if err != nil {
+		return input, pullRequestHash, fmt.Errorf("%s: unable to get hash info", input)
+	}
+
+	return returnedHash, pullRequestHash, nil
+}
+
+// installTt installs selected version of tt.
+func installTt(binDir string, installCtx InstallCtx, distfiles string) error {
+	versionFound := false
+	pullRequestHash := ""
+
 	// Get latest version if it was not specified.
 	ttVersion := installCtx.version
 	if ttVersion == "" {
 		log.Infof("Getting latest tt version...")
+		versions, err := getVersionsFromRepo(installCtx.Local, distfiles, "tt", search.GitRepoTT)
+		if err != nil {
+			return err
+		}
 		if len(versions) == 0 {
 			return fmt.Errorf("no versions were fetched")
 		} else {
 			ttVersion = versions[len(versions)-1].Str
 		}
 	}
 
+	isPullRequest, pullRequestID := util.IsPullRequest(ttVersion)
+
 	// Check that the version exists.
 	// The tag format in tt is vX.Y.Z, but the user can use the X.Y.Z format
 	// and this option needs to be supported.
 	if ttVersion != "master" {
-		versionFound := false
-		for _, ver := range versions {
-			if ttVersion == ver.Str || (ttVersion[0:1] != "v" && "v"+ttVersion == ver.Str) {
-				versionFound = true
-				ttVersion = ver.Str
-				break
+		_, err := version.Parse(ttVersion)
```

Comment:
Am I right, that we try to parse the version line here, even if we already know that it is a pull request?

Reply:
I put a check on the version first, I left the check on the pullrequest for later. I was not too lazy, I wrote a benchmark, and a version.Parse outputs error faster than IsPullRequest returns bool

Reply:
This is no about performance, it is about logic: I've checked that the input is a PR. Why is version parsing not under condition "if not PR"?